### PR TITLE
Updating the `intptr-operators` proposal to include the list of operators.

### PR DESCRIPTION
**Pull request by tannergooding**
FYI. @MadsTorgersen, @gafter

### CODE REVIEW

**Comment by tannergooding** [comment]
Also FYI: @migueldeicaza, since you showed interest in this proposal on the mailing list (in relation to Xamarin)

File: proposals/intptr-operators.md [hunk 1]
```diff
@@ -30,7 +30,245 @@ As such, the language should begin to support these operators to help advance th
...
+```
+
+```C#
+System.UIntPtr operator ~(System.UIntPtr)
+```
+
+#### Cast Operators
+
+```C#
+operator sbyte(System.IntPtr)               // Truncate
```
Comment (gafter) [inline proposals/intptr-operators.md]: Explicit conversions are not specified as methods (if it were, these would need to be `explicit` oeprators), but are rather specified as changes to the spec section on explicit conversions. But this is fine for now.

File: proposals/intptr-operators.md [hunk 2]
```diff
@@ -30,7 +30,245 @@ As such, the language should begin to support these operators to help advance th
...
+operator uint(System.IntPtr)                // Truncate
+operator ulong(System.IntPtr)               // Zero Extend
+
+operator System.IntPtr(int)                 // Sign Extend
+operator System.IntPtr(long)                // Truncate
+
+operator System.IntPtr(uint)                // Sign Extend
+operator System.IntPtr(ulong)               // Truncate
+
+operator System.IntPtr(System.IntPtr)
```
Comment (gafter) [inline proposals/intptr-operators.md]: Should this one return UIntPtr?

File: proposals/intptr-operators.md [hunk 3]
```diff
@@ -30,7 +30,245 @@ As such, the language should begin to support these operators to help advance th
...
+#### Division Operator
+
+```C#
+System.IntPtr operator /(int, System.IntPtr)
+System.IntPtr operator /(System.IntPtr, int)
+System.IntPtr operator /(System.IntPtr, System.IntPtr)
+```
+
+```C#
+System.UIntPtr operator /(uint, USystem.IntPtr)
```
Comment (gafter) [inline proposals/intptr-operators.md]: I think `USystem.IntPtr` should be `System.UIntPtr`

File: proposals/intptr-operators.md (hunk 2 above)
Comment (tannergooding) [inline proposals/intptr-operators.md]: No this was meant to be casting to same type (so `System.IntPtr x = (System.IntPtr)System.IntPtr.Zero`) for example, it (like the unary plus operators) is effectively a no-op.

The ones that return `System.UIntPtr` are just below these.

I did

File: proposals/intptr-operators.md (hunk 3 above)
Comment (tannergooding) [inline proposals/intptr-operators.md]: Yes, will fix.

File: proposals/intptr-operators.md (hunk 1 above)
Comment (tannergooding) [inline proposals/intptr-operators.md]: Yes, I wasn't sure how best to express these, but they were meant to all be `explicit operator`

Comment (tannergooding) [inline proposals/intptr-operators.md]: I have updated to explicitly list these as `explicit operator`.

**Comment by tannergooding** [comment]
I believe all feedback has been addressed here.

**Comment by gafter** [comment]
@MadsTorgersen If you're good with the proposal in this state, please let me know and I'll merge this PR and create a tracking issue.

**Comment by MadsTorgersen** [comment]
Looks good!

**Comment by dgrunwald** [comment]
What's the semantics of these cast operators? The same as if `int`/`long` (depending on platform) was used in place of `IntPtr`? That would mean you get overflow checks based on the current `checked`/`unchecked` context.
While that would make sense, it's a breaking change from the present [crazy behavior](https://github.com/icsharpcode/ILSpy/blob/master/doc/IntPtr.txt#L27) that these casts exhibit. Worse, there might be programs out there that rely on the current behavior for security (overflow checks), so changing the language could introduce security vulnerabilities into existing programs.

**Comment by tannergooding** [comment]
@dgrunwald, there are a few proposals around about `intptr` in general.

The two most likely (at least last I checked) are either:
* A new type
* A new keyword that uses partial erasure under the hood (I am in favor of this approach, personally)

In either case, this would not be a breaking change.

**Pull request merged**